### PR TITLE
[2201.3.x] Add examples for the array langlib

### PR DESCRIPTION
**Pull request by MaryamZi**
## Purpose
$title.

Previous PRs - https://github.com/ballerina-platform/ballerina-lang/pull/38892, https://github.com/ballerina-platform/ballerina-lang/pull/38989

Closes https://github.com/ballerina-platform/ballerina-lang/issues/38907

### CODE REVIEW

File: langlib/lang.array/src/main/ballerina/array.bal [hunk 1]
```diff
@@ -225,6 +324,10 @@ public isolated function lastIndexOf(AnydataType[] arr, AnydataType val, int sta
 
 # Reverses the order of the members of an array.
 #
+# ```ballerina
+# [2, 4, 6, 8, 10].reverse() ⇒ [10,8,6,4,2]
```
Comment (pcnfernando) [inline langlib/lang.array/src/main/ballerina/array.bal:328]: Just to show that its not in descending order
```suggestion
# [2, 4, 12, 8, 10].reverse() ⇒ [10,8,12,4,2]
```

File: langlib/lang.array/src/main/ballerina/array.bal [hunk 2]
```diff
@@ -104,11 +137,9 @@ public isolated function filter(Type[] arr, @isolatedParam function(Type val) re
 # The combining function takes the combined value so far and a member of the array,
 # and returns a new combined value.
 #
-# For example
-# ```
-# reduce([1, 2, 3], function (int total, int n) returns int { return total + n; }, 0)
+# ```ballerina
+# [1, 2, 3].reduce(isolated function (int total, int next) returns int => total + next, 0) ⇒ 6
```
Comment (dulajdilshan) [inline langlib/lang.array/src/main/ballerina/array.bal:141]: `total` or `accTotal` (I'm referring to accumulated total as `accTotal`)?

**Comment by pcnfernando** [comment]
Do we have the same for master?

**Comment by MaryamZi** [comment]
> Do we have the same for master?

Will send once this is approved (that was the original plan since the content may keep changing).

**Review by pcnfernando — APPROVED** [review]

File: langlib/lang.array/src/main/ballerina/array.bal [hunk 3]
```diff
@@ -138,6 +174,16 @@ public isolated function some(Type[] arr, @isolatedParam function(Type val) retu
 
 # Returns a subarray using a start index (inclusive) and an end index (exclusive).
 #
+# ```ballerina
+# int[] numbers = [2, 4, 6, 8, 10, 12];
+#
+# numbers.slice(3) ⇒ [8,10,12]
```
Comment (dulajdilshan) [inline langlib/lang.array/src/main/ballerina/array.bal:180]:
```suggestion
# numbers.slice(3) ⇒ [8, 10, 12]
```

File: langlib/lang.array/src/main/ballerina/array.bal [hunk 4]
```diff
@@ -138,6 +174,16 @@ public isolated function some(Type[] arr, @isolatedParam function(Type val) retu
...
+#
+# numbers.slice(0, 4) ⇒ [2,4,6,8]
```
Comment (dulajdilshan) [inline langlib/lang.array/src/main/ballerina/array.bal:182]:
```suggestion
# numbers.slice(0, 4) ⇒ [2, 4, 6, 8]
```

File: langlib/lang.array/src/main/ballerina/array.bal [hunk 5]
```diff
@@ -190,6 +259,16 @@ public isolated function removeAll((any|error)[] arr) returns () = @java:Method
 #
 # `setLength(arr, 0)` is equivalent to `removeAll(arr)`.
 #
+# ```ballerina
+# int[] numbers = [2, 4, 6, 8];
+#
+# numbers.setLength(2);
+# numbers ⇒ [2,4]
```
Comment (dulajdilshan) [inline langlib/lang.array/src/main/ballerina/array.bal:266]:
```suggestion
# numbers ⇒ [2, 4]
```

File: langlib/lang.array/src/main/ballerina/array.bal [hunk 6]
```diff
@@ -247,6 +350,18 @@ public type OrderedType ()|boolean|int|float|decimal|string|OrderedType[];
 # must be specified.
 # Sorting works the same as with the parameter `sort` clause of query expressions.
 #
+# ```ballerina
+# string[] strings = ["c", "a", "B"];
+#
+# strings.sort() ⇒ ["B","a","c"]
```
Comment (dulajdilshan) [inline langlib/lang.array/src/main/ballerina/array.bal:356]:
```suggestion
# strings.sort() ⇒ ["B", "a", "c"]
```

File: langlib/lang.array/src/main/ballerina/array.bal [hunk 7]
```diff
@@ -247,6 +350,18 @@ public type OrderedType ()|boolean|int|float|decimal|string|OrderedType[];
...
+#
+# strings.sort("descending") ⇒ ["c","a","B"]
+#
+# strings.sort(key = string:toLowerAscii) ⇒ ["a","B","c"]
+#
+# strings.sort("descending", string:toLowerAscii) ⇒ ["c","B","a"]
+# ```
```
Comment (dulajdilshan) [inline langlib/lang.array/src/main/ballerina/array.bal:363]:
```suggestion
# strings.sort("descending") ⇒ ["c", "a", "B"]
#
# strings.sort(key = string:toLowerAscii) ⇒ ["a", "B", "c"]
#
# strings.sort("descending", string:toLowerAscii) ⇒ ["c", "B", "a"]
# ```
```

File: langlib/lang.array/src/main/ballerina/array.bal [hunk 8]
```diff
@@ -274,6 +397,17 @@ public isolated function pop(Type[] arr) returns Type = @java:Method {
 
 # Adds values to the end of an array.
 #
+# ```ballerina
+# int[] numbers = [2];
+#
+# numbers.push(4, 6);
+# numbers ⇒ [2,4,6]
```
Comment (dulajdilshan) [inline langlib/lang.array/src/main/ballerina/array.bal:404]:
```suggestion
# numbers ⇒ [2, 4, 6]
```

File: langlib/lang.array/src/main/ballerina/array.bal [hunk 9]
```diff
@@ -274,6 +397,17 @@ public isolated function pop(Type[] arr) returns Type = @java:Method {
...
+#
+# int[] moreNumbers = [8, 10, 12, 14];
+# numbers.push(...moreNumbers);
+# numbers ⇒ [2,4,6,8,10,12,14]
```
Comment (dulajdilshan) [inline langlib/lang.array/src/main/ballerina/array.bal:408]:
```suggestion
# numbers ⇒ [2, 4, 6, 8, 10, 12, 14]
```

File: langlib/lang.array/src/main/ballerina/array.bal [hunk 10]
```diff
@@ -301,6 +443,17 @@ public isolated function shift(Type[] arr) returns Type = @java:Method {
...
+# ```ballerina
+# int[] numbers = [14];
+#
+# numbers.unshift(10, 12);
+# numbers ⇒ [10,12,14]
+#
+# int[] moreNumbers = [2, 4, 6, 8];
+# numbers.unshift(...moreNumbers);
+# numbers ⇒ [2,4,6,8,10,12,14]
+# ```
```
Comment (dulajdilshan) [inline langlib/lang.array/src/main/ballerina/array.bal:455]:
```suggestion
# numbers.unshift(10, 12);
# numbers ⇒ [10, 12, 14]
#
# int[] moreNumbers = [2, 4, 6, 8];
# numbers.unshift(...moreNumbers);
# numbers ⇒ [2, 4, 6, 8, 10, 12, 14]
# ```
```

File: langlib/lang.array/src/main/ballerina/array.bal [hunk 11]
```diff
@@ -362,6 +537,11 @@ public isolated function fromBase16(string str) returns byte[]|error = @java:Met
 
 # Returns a stream from the given array.
 #
+# ```ballerina
+# stream<string> strm = ["a", "b", "c", "d"].toStream();
+# strm.next() ⇒ {"value":"a"}
```
Comment (dulajdilshan) [inline langlib/lang.array/src/main/ballerina/array.bal:542]:
```suggestion
# strm.next() ⇒ {"value": "a"}
```

**Review by dulajdilshan — COMMENTED** [review]
added dew comments please look

File: langlib/lang.array/src/main/ballerina/array.bal [hunk 12]
```diff
@@ -46,8 +50,15 @@ public isolated function length((any|error)[] arr) returns int = @java:Method {
 
 # Returns an iterator over an array.
 #
+# ```ballerina
+# object {
+#     public isolated function next() returns record {|int value;|}?;
+# } iterator = [2, 4, 6, 8].iterator();
+# iterator.next() ⇒ {"value":2}
```
[the remark this answers is not in the document]
Comment (MaryamZi) [reply langlib/lang.array/src/main/ballerina/array.bal:57]: All of this is the `toString` representation, which is the output with `io:println` and (often) shell. So it is intentional.

File: langlib/lang.array/src/main/ballerina/array.bal (hunk 2 above)
Comment (MaryamZi) [reply langlib/lang.array/src/main/ballerina/array.bal:141]: Doesn't total indicate that? Why do we need to add accumulated?

File: langlib/lang.array/src/main/ballerina/array.bal (hunk 12 above)
[the remark this answers is not in the document]
Comment (dulajdilshan) [reply langlib/lang.array/src/main/ballerina/array.bal:57]: Ah okay. Let's ignore my suggestions

**Review by dulajdilshan — APPROVED** [review]
LGTM

File: langlib/lang.array/src/main/ballerina/array.bal (hunk 2 above)
Comment (dulajdilshan) [reply langlib/lang.array/src/main/ballerina/array.bal:141]: Doesn't the user get a confusing idea? (total and accumulated total are two different things, no?

Please check this too: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce

BTW If we want to focus on showing minimal effort, I'm okay using `total`.

Comment (MaryamZi) [reply langlib/lang.array/src/main/ballerina/array.bal:141]: I'm not sure that `total` and `accumulatedTotal` are two different things, depends on the context, right? Here we are "updating" `total` with the accumulated value so far.

> Please check this too: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce

Doesn't this use `accumulator`? I think `total` OR `accumulator` make sense, but `accumulatedTotal` doesn't add much value compared to `total`.

Comment (dulajdilshan) [reply langlib/lang.array/src/main/ballerina/array.bal:141]: hmm... Yeah. `total` could be better

**Pull request merged**